sklearn/experimental/_column_transformer.py

```diff
@@ -0,0 +1,282 @@
+            one result is a sparse matrix, everything will be converted to
+            sparse matrices.
+        """
+        try:
+            return super(ColumnTransformer, self).transform(X)
+        except ValueError as e:
+            if "Got X with X.ndim=1. Reshape your data" in str(e):
+                raise ValueError(_ERR_MSG_1DCOLUMN)
+            else:
+                raise e
+
+
+def _get_column(X, key):
+    """
+    Get feature column(s) from input data X.
+
+    Supported input types (X): numpy arrays, sparse arrays and dataframes
+
+    Supported key types (key):
+    - scalar: output is 1D
+    - lists, slices, boolean masks: output is 2D
+
+    Supported key data types:
+
+    - integer or boolean mask (positional):
+        - supported for (sparse) arrays or dataframes
+    - string (key-based):
+        - only supported for dataframes
+        - So no keys other than strings are allowed (while in principle you
+          can use any hashable object as key).
+
+    """
+    if X is None:
+        return X
+
+    # check whether we have string column names or integers
+    if (isinstance(key, int)
+            or (isinstance(key, list)
+                and all(isinstance(col, int) for col in key))
+            or (isinstance(key, slice)
+                and isinstance(key.start, (int, type(None)))
+                and isinstance(key.stop, (int, type(None))))):
+        column_names = False
+    elif (isinstance(key, six.string_types)
+            or (isinstance(key, list)
+                and all(isinstance(col, six.string_types) for col in key))
+            or (isinstance(key, slice)
+                and isinstance(key.start, (six.string_types, type(None)))
+                and isinstance(key.stop, (six.string_types, type(None))))):
+        column_names = True
+    elif hasattr(key, 'dtype') and np.issubdtype(key.dtype, np.bool):
+        # boolean mask
+        column_names = False
+        if hasattr(X, 'loc'):
+            # pandas boolean masks don't work with iloc, so take loc path
+            column_names = True
+    else:
+        raise ValueError("No valid specification of the columns. Only a "
+                         "scalar, list or slice of all integers or all "
+                         "strings, or boolean mask is allowed")
+
+    if column_names:
+        if hasattr(X, 'loc'):
+            # pandas dataframes
+            return X.loc[:, key]
+        else:
+            raise ValueError("Specifying the columns using strings is only "
+                             "supported for pandas DataFrames")
+    else:
+        if hasattr(X, 'iloc'):
+            # pandas dataframes
+            return X.iloc[:, key]
+        else:
+            # numpy arrays, sparse arrays
+            return X[:, key]
+
+
+def _get_transformer_list(estimators):
+    """
+    Construct (name, trans, column) tuples from dictionary
+
+    """
+    transformers = estimators.keys()
```

Comment:
The iteration order here is not deterministic. I think we need it to be.

Making it deterministic will be hard if we base it on the columns transformed. For practicality we'll need to sort the names.